### ISSUE TITLE
No diagnostics after the client responds to `workspace/configuration`

### ISSUE THREAD

**Issue by bstaletic** [rust-analyzer/rust-analyzer#5464]
Because some other server required it, ycmd is doing the following:

1. Send `initialize` request and wait for response.
2. Send `initialized` notification.
3. Send `workspace/didChangeConfiguration` notification.

This triggers `workspace/configuration` request, to which, provided no user configuration, ycmd will respond with `[null]`. This, however, results in rust-analyzer sending empty diagnostics until the next `didSave` notification.

Here's the full log of exchanged messages:

```
2020-07-20 21:46:58,432 - DEBUG - TX: Sending message: b'Content-Length: 1052\r\n\r\n{"id":1,"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"experimental":{"statusNotification":true},"textDocument":{"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"completion":{"completionItem":{"documentationFormat":["plaintext","markdown"]},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"hover":{"contentFormat":["plaintext","markdown"]},"signatureHelp":{"signatureInformation":{"documentationFormat":["plaintext","markdown"],"parameterInformation":{"labelOffsetSupport":false}}},"synchronization":{"didSave":true}},"workspace":{"configuration":true}},"initializationOptions":{},"processId":12969,"rootPath":"/home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common","rootUri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common"}}'
2020-07-20 21:46:58,433 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":1,"result":{"capabilities":{"callHierarchyProvider":true,"codeActionProvider":{"codeActionKinds":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite"]},"codeLensProvider":{"resolveProvider":true},"completionProvider":{"triggerCharacters":[":","."]},"definitionProvider":true,"documentFormattingProvider":true,"documentHighlightProvider":true,"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"=","moreTriggerCharacter":[".",">"]},"documentSymbolProvider":true,"experimental":{"joinLines":true,"onEnter":true,"parentModule":true,"runnables":{"kinds":["cargo"]},"ssr":true},"foldingRangeProvider":true,"hoverProvider":true,"implementationProvider":true,"referencesProvider":true,"renameProvider":{"prepareProvider":true},"selectionRangeProvider":true,"semanticTokensProvider":{"documentProvider":true,"legend":{"tokenModifiers":["documentation","declaration","definition","static","abstract","deprecated","readonly","constant","controlFlow","injected","mutable","unsafe","attribute"],"tokenTypes":["comment","keyword","string","number","regexp","operator","namespace","type","struct","class","interface","enum","typeParameter","function","member","property","macro","variable","parameter","label","attribute","boolean","builtinType","enumMember","escapeSequence","formatSpecifier","generic","lifetime","selfKeyword","typeAlias","union","unresolvedReference"]},"rangeProvider":true},"signatureHelpProvider":{"triggerCharacters":["(",","]},"textDocumentSync":{"change":2,"openClose":true,"save":{}},"typeDefinitionProvider":true,"workspaceSymbolProvider":true},"serverInfo":{"name":"rust-analyzer","version":"f5a4a4b"}}}'
2020-07-20 21:46:58,435 - DEBUG - TX: Sending notification: b'Content-Length: 52\r\n\r\n{"jsonrpc":"2.0","method":"initialized","params":{}}'
2020-07-20 21:46:58,435 - DEBUG - TX: Sending notification: b'Content-Length: 86\r\n\r\n{"jsonrpc":"2.0","method":"workspace/didChangeConfiguration","params":{"settings":{}}}'
2020-07-20 21:46:58,436 - DEBUG - TX: Sending notification: b'Content-Length: 555\r\n\r\n{"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"rust","text":"mod test;\\n\\nuse test::*;\\n\\nstruct Earth {}\\nstruct Mars {}\\ntrait Atmosphere {}\\nimpl Atmosphere for Earth {}\\nimpl Atmosphere for Mars {}\\n\\nfn main() {\\n    create_universe();\\n    let builder = Builder {};\\n    builder.build_\\n}\\n\\nfn             format_test() {\\n\\n    let\\n        a : \\ni32 = 5;\\n}\\n","uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs","version":1}}}'
2020-07-20 21:46:58,437 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":0,"method":"client/registerCapability","params":{"registrations":[{"id":"textDocument/didSave","method":"textDocument/didSave","registerOptions":{"documentSelector":[{"pattern":"**/*.rs"},{"pattern":"**/Cargo.toml"},{"pattern":"**/Cargo.lock"}]}}]}}'
2020-07-20 21:46:58,437 - DEBUG - TX: Sending response: b'Content-Length: 38\r\n\r\n{"id":0,"jsonrpc":"2.0","result":null}'
2020-07-20 21:46:58,437 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":1,"method":"workspace/configuration","params":{"items":[{"section":"rust-analyzer"}]}}'
2020-07-20 21:46:58,437 - DEBUG - TX: Sending response: b'Content-Length: 40\r\n\r\n{"id":1,"jsonrpc":"2.0","result":[null]}'
2020-07-20 21:46:58,440 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"rust-analyzer/status","params":"invalid"}'
2020-07-20 21:46:58,441 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[],"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs"}}'
2020-07-20 21:46:58,514 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"rust-analyzer/status","params":"loading"}'
2020-07-20 21:46:58,602 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"rust-analyzer/status","params":"ready"}'
2020-07-20 21:47:02,111 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[],"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs"}}'
2020-07-20 21:47:07,142 - DEBUG - TX: Sending notification: b'Content-Length: 218\r\n\r\n{"jsonrpc":"2.0","method":"textDocument/didSave","params":{"textDocument":{"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs","version":1}}}'
2020-07-20 21:47:07,673 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"code":"E0609","message":"no field `build_` on type `test::Builder`\\nunknown field","range":{"end":{"character":18,"line":13},"start":{"character":12,"line":13}},"severity":1,"source":"rustc"}],"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs"}}'
```

On request, I can provide rust-analyzer's log too, just tell me at what verbosity level.

Another thing to note in the above log is that the first `rust-analyzer/status` is `invalid`, then `loading` and then `ready`. What's up with the `invalid` one?

Finally, contrast the above log with the following one, where ycmd returns an error response on `worksapce/configuration`.

```
2020-07-20 21:53:31,189 - DEBUG - TX: Sending message: b'Content-Length: 1232\r\n\r\n{"id":1,"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"experimental":{"statusNotification":true},"textDocument":{"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"completion":{"completionItem":{"documentationFormat":["plaintext","markdown"]},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"hover":{"contentFormat":["plaintext","markdown"]},"signatureHelp":{"signatureInformation":{"documentationFormat":["plaintext","markdown"],"parameterInformation":{"labelOffsetSupport":false}}},"synchronization":{"didSave":true}},"workspace":{"applyEdit":true,"didChangeWatchedFiles":{"dynamicRegistration":true},"documentChanges":true,"symbol":{"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}}}},"initializationOptions":{},"processId":14895,"rootPath":"/home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common","rootUri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common"}}'
2020-07-20 21:53:31,190 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":1,"result":{"capabilities":{"callHierarchyProvider":true,"codeActionProvider":{"codeActionKinds":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite"]},"codeLensProvider":{"resolveProvider":true},"completionProvider":{"triggerCharacters":[":","."]},"definitionProvider":true,"documentFormattingProvider":true,"documentHighlightProvider":true,"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"=","moreTriggerCharacter":[".",">"]},"documentSymbolProvider":true,"experimental":{"joinLines":true,"onEnter":true,"parentModule":true,"runnables":{"kinds":["cargo"]},"ssr":true},"foldingRangeProvider":true,"hoverProvider":true,"implementationProvider":true,"referencesProvider":true,"renameProvider":{"prepareProvider":true},"selectionRangeProvider":true,"semanticTokensProvider":{"documentProvider":true,"legend":{"tokenModifiers":["documentation","declaration","definition","static","abstract","deprecated","readonly","constant","controlFlow","injected","mutable","unsafe","attribute"],"tokenTypes":["comment","keyword","string","number","regexp","operator","namespace","type","struct","class","interface","enum","typeParameter","function","member","property","macro","variable","parameter","label","attribute","boolean","builtinType","enumMember","escapeSequence","formatSpecifier","generic","lifetime","selfKeyword","typeAlias","union","unresolvedReference"]},"rangeProvider":true},"signatureHelpProvider":{"triggerCharacters":["(",","]},"textDocumentSync":{"change":2,"openClose":true,"save":{}},"typeDefinitionProvider":true,"workspaceSymbolProvider":true},"serverInfo":{"name":"rust-analyzer","version":"f5a4a4b"}}}'
2020-07-20 21:53:31,191 - DEBUG - TX: Sending notification: b'Content-Length: 52\r\n\r\n{"jsonrpc":"2.0","method":"initialized","params":{}}'
2020-07-20 21:53:31,191 - DEBUG - TX: Sending notification: b'Content-Length: 86\r\n\r\n{"jsonrpc":"2.0","method":"workspace/didChangeConfiguration","params":{"settings":{}}}'
2020-07-20 21:53:31,192 - DEBUG - Refreshing file /home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs: State is Open/action Open
2020-07-20 21:53:31,192 - DEBUG - TX: Sending notification: b'Content-Length: 555\r\n\r\n{"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"rust","text":"mod test;\\n\\nuse test::*;\\n\\nstruct Earth {}\\nstruct Mars {}\\ntrait Atmosphere {}\\nimpl Atmosphere for Earth {}\\nimpl Atmosphere for Mars {}\\n\\nfn main() {\\n    create_universe();\\n    let builder = Builder {};\\n    builder.build_\\n}\\n\\nfn             format_test() {\\n\\n    let\\n        a : \\ni32 = 5;\\n}\\n","uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs","version":1}}}'
2020-07-20 21:53:31,193 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":0,"method":"client/registerCapability","params":{"registrations":[{"id":"textDocument/didSave","method":"textDocument/didSave","registerOptions":{"documentSelector":[{"pattern":"**/*.rs"},{"pattern":"**/Cargo.toml"},{"pattern":"**/Cargo.lock"}]}}]}}'
2020-07-20 21:53:31,193 - DEBUG - TX: Sending response: b'Content-Length: 38\r\n\r\n{"id":0,"jsonrpc":"2.0","result":null}'
2020-07-20 21:53:31,193 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":1,"method":"workspace/configuration","params":{"items":[{"section":"rust-analyzer"}]}}'
2020-07-20 21:53:31,193 - DEBUG - TX: Sending response: b'Content-Length: 77\r\n\r\n{"error":{"code":-32601,"message":"Method not found"},"id":1,"jsonrpc":"2.0"}'
2020-07-20 21:53:31,271 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"rust-analyzer/status","params":"loading"}'
2020-07-20 21:53:31,357 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"rust-analyzer/status","params":"ready"}'
2020-07-20 21:53:31,430 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"code":"E0609","message":"no field `build_` on type `test::Builder`\\nunknown field","range":{"end":{"character":18,"line":13},"start":{"character":12,"line":13}},"severity":1,"source":"rustc"}],"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs"}}'
2020-07-20 21:53:34,870 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"code":"E0609","message":"no field `build_` on type `test::Builder`\\nunknown field","range":{"end":{"character":18,"line":13},"start":{"character":12,"line":13}},"severity":1,"source":"rustc"}],"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs"}}'
```

Differences between the two:

1. There's no `invalid` status at any point.
2. Diagnostics arrive as expected.

Finally, I'm testing with this file/project: https://github.com/ycm-core/ycmd/blob/master/ycmd/tests/rust/testdata/common/src/main.rs

**Comment by kjeremy** [comment]
Why does the second example throw an error? Is that just for your testing?

**Comment by bstaletic** [comment]
Because we currently do not support `workspace/configuration` and do not declare the related capability. Therefore the polite thing to do is return an error response with the code `MethodNotImplemented`, as opposed to just not responding to that request.

**Comment by kjeremy** [comment]
I think what's happening is that the `[null]` returned from `workspace/configuration` is turning off the diagnostics and the flycheck gets triggered on the next save.

https://github.com/rust-analyzer/rust-analyzer/blob/76c1fac9c3ed9fa3f784e5cdde7dd770c9fac6c1/crates/rust-analyzer/src/main_loop.rs#L451-L484

https://github.com/rust-analyzer/rust-analyzer/blob/76c1fac9c3ed9fa3f784e5cdde7dd770c9fac6c1/crates/rust-analyzer/src/reload.rs#L19-L30



Can you turn on [debug logging](https://github.com/rust-analyzer/rust-analyzer/tree/master/docs/dev#logging) so we can see what "config update response:" is?

**Comment by kjeremy** [comment]
By the same token [InitializationOptions](https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#initializationoptions) having an empty value may also be problematic. You may not want to send that down at all.

**Comment by bstaletic** [comment]
> I think what's happening is that the `[null]` returned from `workspace/configuration` is turning off the diagnostics and the flycheck gets triggered on the next save.

I don't know if it makes any difference, but I can also respond with `[{}]` and have the same result.

> Can you turn on debug logging so we can see what "config update response:" is?

The above logs contain the LSP messages, just from the client's (ycmd's) perspective. The RA's log, with `RA_LOG='rust_analyzer=debug'` is here: https://gist.github.com/bstaletic/70f4ecb45616166b1dfbad44d1972226

**Comment by bstaletic** [comment]
> By the same token [InitializationOptions](https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#initializationoptions) having an empty value may also be problematic. You may not want to send that down at all.

You mean I shouldn't be sending `{}` in `initializationOptions`? But it's very much possible I have nothing else to send.

**Comment by kjeremy** [comment]
I'm saying if `initializationOptions` doesn't have the `rust-analyzer` section leave it out of `InitializeParams` entirely (since `initializationOptions` is optional). It looks like it will overwrite the default config:

https://github.com/rust-analyzer/rust-analyzer/blob/2c67ca0146c215df908de0066b124d63d2ba8fe7/crates/rust-analyzer/src/bin/main.rs#L124-L126

I'll try to verify this in a bit.

**Comment by kjeremy** [comment]
Also you are right in that you should be able to send `[null]` in `workspace/configuration`. I'll see if I can fix that.

**Comment by bstaletic** [comment]
> I'm saying if `initializationOptions` doesn't have the `rust-analyzer` section leave it out of `InitializeParams` entirely (since `initializationOptions` is optional).

Okay, now I understand. However, I'd call this problematic. Right now everything regarding `initializationOptions` is completely generic. 

https://github.com/ycm-core/ycmd/blob/master/ycmd/completers/language_server/language_server_protocol.py#L269

https://github.com/ycm-core/ycmd/blob/master/ycmd/completers/language_server/language_server_completer.py#L1946

Implementing "if rust and rust-analyzer not in settings" is VERY server specific.

> Also you are right in that you should be able to send `[null]` in `workspace/configuration`. I'll see if I can fix that.

I know @matklad said that it should work.

**Comment by kjeremy** [comment]
Can you see if https://github.com/kjeremy/rust-analyzer/tree/null-configuration helps?

**Comment by kjeremy** [comment]
I can reproduce this if I send a `null` for `initializeOptions`. I THINK it should be safe to filter that out.

**Comment by bstaletic** [comment]
> Can you see if https://github.com/kjeremy/rust-analyzer/tree/null-configuration helps?

With that branch I didn't get *any* diagnostics at all, even after `didSave`.

One more question regarding `initializationOptions`. Should the value be `{ 'rust-analyzer': <settings> }` or just `<settings>`? From what you're telling me and what @matklad told me before, it should be `{ 'rust-analyzer': <settings> }`. However, [this](https://github.com/rust-analyzer/rust-analyzer/blob/f51b0cfdd6c23dd57a0a11154179730171c0425d/editors/code/src/client.ts#L43) seems to imply that it should just be `<settings>`.

> I'm saying if initializationOptions doesn't have the rust-analyzer section leave it out of InitializeParams entirely (since initializationOptions is optional).

Now that I think about it, ycmd could omit `initializationOptions` if the user did not provide any. So it shouldn't be that much of a problem. And if user supplies options that are invalid, that's on user.

**Comment by kjeremy** [comment]
I just force pushed a new version that should tackle both `workspace/configuration` and `initializationOptions` being `null`

**Comment by kjeremy** [comment]
> One more question regarding `initializationOptions`. Should the value be `{ 'rust-analyzer': <settings> }` or just `<settings>`? From what you're telling me and what @matklad told me before, it should be `{ 'rust-analyzer': <settings> }`. However, [this](https://github.com/rust-analyzer/rust-analyzer/blob/f51b0cfdd6c23dd57a0a11154179730171c0425d/editors/code/src/client.ts#L43) seems to imply that it should just be `<settings>`

`initializationOptions` should be `<settings>`. The branch I posted should allow you to send down a `null` value for `<settings>` and get the defaults.

**Comment by bstaletic** [comment]
At commit `adc4bb54062bd0bdff8bfaecce238990f4225bac`, still not getting any diagnostics.

Here are the LSP messages:

```
2020-07-20 23:49:33,494 - DEBUG - TX: Sending message: b'Content-Length: 1052\r\n\r\n{"id":1,"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"experimental":{"statusNotification":true},"textDocument":{"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"completion":{"completionItem":{"documentationFormat":["plaintext","markdown"]},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"hover":{"contentFormat":["plaintext","markdown"]},"signatureHelp":{"signatureInformation":{"documentationFormat":["plaintext","markdown"],"parameterInformation":{"labelOffsetSupport":false}}},"synchronization":{"didSave":true}},"workspace":{"configuration":true}},"initializationOptions":{},"processId":25600,"rootPath":"/home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common","rootUri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common"}}'
2020-07-20 23:49:33,499 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":1,"result":{"capabilities":{"callHierarchyProvider":true,"codeActionProvider":{"codeActionKinds":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite"]},"codeLensProvider":{"resolveProvider":true},"completionProvider":{"triggerCharacters":[":","."]},"definitionProvider":true,"documentFormattingProvider":true,"documentHighlightProvider":true,"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"=","moreTriggerCharacter":[".",">"]},"documentSymbolProvider":true,"experimental":{"joinLines":true,"onEnter":true,"parentModule":true,"runnables":{"kinds":["cargo"]},"ssr":true},"foldingRangeProvider":true,"hoverProvider":true,"implementationProvider":true,"referencesProvider":true,"renameProvider":{"prepareProvider":true},"selectionRangeProvider":true,"semanticTokensProvider":{"documentProvider":true,"legend":{"tokenModifiers":["documentation","declaration","definition","static","abstract","deprecated","readonly","constant","controlFlow","injected","mutable","unsafe","attribute"],"tokenTypes":["comment","keyword","string","number","regexp","operator","namespace","type","struct","class","interface","enum","typeParameter","function","member","property","macro","variable","parameter","label","attribute","boolean","builtinType","enumMember","escapeSequence","formatSpecifier","generic","lifetime","punctuation","selfKeyword","typeAlias","union","unresolvedReference"]},"rangeProvider":true},"signatureHelpProvider":{"triggerCharacters":["(",","]},"textDocumentSync":{"change":2,"openClose":true,"save":{}},"typeDefinitionProvider":true,"workspaceSymbolProvider":true},"serverInfo":{"name":"rust-analyzer","version":"adc4bb5"}}}'
2020-07-20 23:49:33,501 - DEBUG - TX: Sending notification: b'Content-Length: 52\r\n\r\n{"jsonrpc":"2.0","method":"initialized","params":{}}'
2020-07-20 23:49:33,501 - DEBUG - TX: Sending notification: b'Content-Length: 86\r\n\r\n{"jsonrpc":"2.0","method":"workspace/didChangeConfiguration","params":{"settings":{}}}'
2020-07-20 23:49:33,502 - DEBUG - TX: Sending notification: b'Content-Length: 555\r\n\r\n{"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"rust","text":"mod test;\\n\\nuse test::*;\\n\\nstruct Earth {}\\nstruct Mars {}\\ntrait Atmosphere {}\\nimpl Atmosphere for Earth {}\\nimpl Atmosphere for Mars {}\\n\\nfn main() {\\n    create_universe();\\n    let builder = Builder {};\\n    builder.build_\\n}\\n\\nfn             format_test() {\\n\\n    let\\n        a : \\ni32 = 5;\\n}\\n","uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs","version":1}}}'
2020-07-20 23:49:33,517 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":0,"method":"client/registerCapability","params":{"registrations":[{"id":"textDocument/didSave","method":"textDocument/didSave","registerOptions":{"documentSelector":[{"pattern":"**/*.rs"},{"pattern":"**/Cargo.toml"},{"pattern":"**/Cargo.lock"}],"includeText":false}}]}}'
2020-07-20 23:49:33,517 - DEBUG - TX: Sending response: b'Content-Length: 38\r\n\r\n{"id":0,"jsonrpc":"2.0","result":null}'
2020-07-20 23:49:33,518 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":1,"method":"workspace/configuration","params":{"items":[{"section":"rust-analyzer"}]}}'
2020-07-20 23:49:33,518 - DEBUG - TX: Sending response: b'Content-Length: 40\r\n\r\n{"id":1,"jsonrpc":"2.0","result":[null]}'
2020-07-20 23:49:33,581 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","id":2,"method":"client/registerCapability","params":{"registrations":[{"id":"workspace/didChangeWatchedFiles","method":"workspace/didChangeWatchedFiles","registerOptions":{"watchers":[{"globPattern":"/home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/**/*.rs"}]}}]}}'
2020-07-20 23:49:33,588 - DEBUG - TX: Sending response: b'Content-Length: 38\r\n\r\n{"id":2,"jsonrpc":"2.0","result":null}'
2020-07-20 23:49:33,643 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"rust-analyzer/status","params":"loading"}'
2020-07-20 23:49:33,667 - DEBUG - RX: Received message: b'{"jsonrpc":"2.0","method":"rust-analyzer/status","params":"ready"}'
2020-07-20 23:49:53,109 - DEBUG - TX: Sending notification: b'Content-Length: 218\r\n\r\n{"jsonrpc":"2.0","method":"textDocument/didSave","params":{"textDocument":{"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs","version":1}}}'
2020-07-20 23:49:53,605 - DEBUG - TX: Sending notification: b'Content-Length: 223\r\n\r\n{"jsonrpc":"2.0","method":"workspace/didChangeWatchedFiles","params":{"changes":[{"type":3,"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs"}]}}'
2020-07-20 23:49:53,606 - DEBUG - TX: Sending notification: b'Content-Length: 223\r\n\r\n{"jsonrpc":"2.0","method":"workspace/didChangeWatchedFiles","params":{"changes":[{"type":1,"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs"}]}}'
2020-07-20 23:49:53,606 - DEBUG - TX: Sending notification: b'Content-Length: 223\r\n\r\n{"jsonrpc":"2.0","method":"workspace/didChangeWatchedFiles","params":{"changes":[{"type":2,"uri":"file:///home/bstaletic/.vim/pack/bundle/start/YouCompleteMe/third_party/ycmd/ycmd/tests/rust/testdata/common/src/main.rs"}]}}'
```

And the RA log: https://gist.github.com/bstaletic/70f4ecb45616166b1dfbad44d1972226

**Comment by kjeremy** [comment]
Okay try again (a32dd9c47) should protect against empty values too.

**Comment by bstaletic** [comment]
Compiling.

I almost forgot one thing. Thanks a lot for the instant replies.

**Comment by bstaletic** [comment]
> Compiling.

Done. Works like a charm. Again, ycmd can probably switch to not sending `initializationOptions` when it isn't provided by the user. The `[null]` in the `workspace/configuration` response is more important.

Once again, thanks for working on this.
